CURLOPT_MAXLIFETIME_CONN: set the maximum allowed lifetime for connection reuse

jeffreytolar opened the pull request:
... and close connections that are too old instead of reusing them.

By default, this behavior is disabled.

Bug: https://curl.se/mail/lib-2021-09/0058.html

- - -

This was mostly done by searching for `CURLOPT_MAXAGE_CONN` and adding
corresponding bits for `CURLOPT_MAXLIFETIME_CONN`.

jeffreytolar commented on lib/easyoptions.c:
```diff
@@ -354,6 +355,6 @@ struct curl_easyoption Curl_easyopts[] = {
  */
 int Curl_easyopts_check(void)
 {
-  return ((CURLOPT_LASTENTRY%10000) != (310 + 1));
+  return (CURLOPT_LASTENTRY != (311 + 1));
```
This came when regenerating the file via `optiontable.pl`; if the `%10000` is important, I can add it back (and/or include a commit to add it in the template in `optiontable.pl`)

bagder replied:
It needs the `%10000` there in one way or another, because the last number will be the enum number used on the line above `+1`, and since the number above might be in a numerical range above 10000, doing the logic without the remainder operator could evaluate wrong.

jeffreytolar replied:
Ah, that makes sense now after looking at https://github.com/curl/curl/blob/e5d5e288dd62bf64bdf336152c2391191bea66b7/include/curl/curl.h#L1031-L1041

I'll push a commit to fix that, hopefully later tonight (in a few hours)

bagder commented on docs/libcurl/opts/CURLOPT_MAXLIFETIME_CONN.3:
```diff
@@ -0,0 +1,66 @@
+if(curl) {
+  curl_easy_setopt(curl, CURLOPT_URL, "https://example.com");
+
+  /* only allow each connection to be used for 30 seconds */
```
s/used/reused/ maybe? To stress that this option limits the reuse time, not the use time.

bagder commented on docs/libcurl/opts/CURLOPT_MAXLIFETIME_CONN.3:
```diff
@@ -0,0 +1,66 @@
+seconds, since the creation of the connection, that you allow an existing
+connection to have to be considered for reuse for this request.
+
+The "connection cache" that holds previously used connections. When a new
```
This sentence seems odd. How about:

"libcurl features a connection cache that holds previously used connections."

bagder commented on lib/url.c:
```diff
@@ -962,21 +963,36 @@ socks_proxy_info_matches(const struct proxy_info *data,
-   be subject for reuse. The success rate is just too low after this. */
+/* A connection has to have been idle for a shorter time than 'maxage_conn', or
+   created less than 'maxlifetime_conn' ago, to be subject for reuse. The
+   success rate is just too low after this. */
```
The success rate comment refers to the idle time part, shouldn't this then make the comment instead say:
~~~c
/* A connection has to have been idle for a shorter time than
  'maxage_conn' (the success rate is just too low after this) or
  created less than 'maxlifetime_conn' ago, to be subject for reuse. */

jeffreytolar commented:
Rebased on current `master` to resolve conflicts with the new options; mostly it was `s/311/314/` to update the option enum value. I also removed 1a5e740 (the commit to update `optiontable.pl`) due to f53b89314ddbfe2b3b203a85c2181868c6151363

bagder commented:
Thanks!
